File: python/src/e_external_calls.py
"""
Use pytest to test each case (even exceptions)
You can use Faker to generate random data
You can use Pytest fixtures to 'inject' data/behavior in your tests
You will need to mock function httpx.get
https://pytest-mock.readthedocs.io/en/latest/usage.html
"""
from dataclasses import dataclass

import httpx


@dataclass
class Cat:
    name: str
    breed: str

    def get_image(self, api_key: str) -> str:
        response = httpx.get(
            f"https://api.thecatapi.com/v1/images/search?breed_ids={self.breed}&api_key={api_key}"
        )
        if response.status_code != httpx.codes.OK:
            response.raise_for_status()

        return response.json()[0].get("url")

File: javascript/src/e-external-calls.ts
/**
 * Use jest to test each case (even exceptions)
 * You can use Faker.js to generate random data
 * You will need to mock function httpx.get
 * https://jestjs.io/docs/jest-object#jestspyonobject-methodname
 * https://jestjs.io/docs/mock-function-api#mockfnmockresolvedvaluevalue
 */

export class Cat {
  name: string;
  breed: string;

  constructor(name: string, breed: string) {
    this.name = name;
    this.breed = breed;
  }

  async getImage(apiKey: string): Promise<string> {
    const response = await fetch(
      `https://api.thecatapi.com/v1/images/search?breed_ids=${this.breed}&api_key=${apiKey}`
    );

    if (response.ok) {
      const json = await response.json();
      return json[0]['url'];
    } else {
      throw new Error(`HTTP error ${response.status}`);
    }
  }
}
